lib/kcerts/ssh.go

```diff
@@ -0,0 +1,199 @@
+// For each entry, it adds an additional line and does not concatenate
+func AddSSHCAToClient(publicKey ssh.PublicKey, hosts []string, sshDir string) error {
+	caPublic := string(ssh.MarshalAuthorizedKey(publicKey))
+	knownHosts := filepath.Join(sshDir, KnownHosts)
+	knownHostsFile, err := os.OpenFile(knownHosts, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0644)
+	defer knownHostsFile.Close()
+	if err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("could not create known_hosts file: %w", err)
+		}
+		return err
+	}
+	existingKnownHostsContent, err := ioutil.ReadAll(knownHostsFile)
+	if err != nil {
+		return fmt.Errorf("error reading %s: %w", knownHosts, err)
+	}
+	for _, dns := range hosts {
+		// caPublic terminates with a '\n', added by ssh.MarshalAuthorizedKey
+		publicFormat := fmt.Sprintf("%s %s %s", CAPrefix, dns, caPublic)
+		if strings.Contains(string(existingKnownHostsContent), publicFormat) {
+			continue
+		}
+		_, err = knownHostsFile.WriteString(publicFormat)
+		if err != nil {
+			return fmt.Errorf("could not add key %s to known_hosts file: %w", publicFormat, err)
+		}
+	}
+	return nil
+}
+
+type SSHAgent struct {
+	PID    int    `json:"pid"`
+	Socket string `json:"sock"`
+	// failed is marked in WriteToCache
+	failed bool
+}
+
+func (a SSHAgent) Valid() bool {
+	conn, err := net.Dial("unix", a.Socket)
+	defer conn.Close()
+	return err == nil
+}
+
+// CloseIfNotCached will kill the agent if it failed to write to cache. Otherwise is a NoOp
+func (a *SSHAgent) CloseIfNotCached() error {
+	if a.failed {
+		cmd := exec.Command("kill", "-HUP", strconv.Itoa(a.PID))
+		cmd.Stdout = os.Stdout
+		return cmd.Run()
+	}
+	return nil
+}
+
+// FindSSHAgent Will start the ssh agent in the interactive terminal if it isn't present already as an environment variable
+// Currently only outputs the env and does not persist it across terminals
+func FindSSHAgent(store cache.Store, logger *klog.Logger) (*SSHAgent, error) {
+	// TODO(adam): sniff out dead agents and sneak pids
+	agent := FindSSHAgentFromEnv()
+	if agent != nil && agent.Valid() {
+		return agent, nil
+	}
+	agent, err := FetchSSHAgentFromCache(store)
+	if err != nil {
+		logger.Warn(err.Error())
+		if errors.Is(SSHAgentCacheInvalid, err) {
+			if err := DeleteSSHCache(store); err != nil {
+				logger.Warnf("could not delete cache %w", err)
+			}
+		}
+	}
+	if agent != nil && agent.Valid() {
+		return agent, nil
+	}
+	newAgent, err := CreateNewSSHAgent()
+	if err != nil {
+		return nil, err
+	}
+	defer logger.Info(WriteAgentToCache(store, newAgent))
+	return newAgent, nil
+}
+
+// FindSSHAgentFromEnv
+func FindSSHAgentFromEnv() *SSHAgent {
+	envSSHSock := os.Getenv("SSH_AUTH_SOCK")
+	envSSHPID := os.Getenv("SSH_AGENT_PID")
+	if envSSHSock != "" && envSSHPID != "" {
+		pid, err := strconv.Atoi(envSSHPID)
+		if err != nil {
+			return nil
+		}
+		return &SSHAgent{PID: pid, Socket: envSSHSock}
```

Comment:
`&SSHAgent{Close: func () {}, PID: pid, Socket: envSSHSock}`

(nothing to do if we try to close this agent, does not belong to us)